text/3334-niche.md

```diff
@@ -0,0 +1,348 @@
+errors or missing data) within the same bits.
+
+The Rust compiler already provides support for this via "niche" optimizations,
+and various types providing guarantees of such optimizations, including
+references, `bool`, `char`, and the `NonZero` family of types. However, Rust
+does not provide any stable means of defining new types with niches, reserving
+this mechanism for the standard library. This puts pressure on the standard
+library to provide additional families of types with niches, while preventing
+the broader crate ecosystem from experimenting with such types.
+
+Past efforts to define a stable niche mechanism stalled out due to scope creep:
+alignment niches, null-page niches, multiple niches, structures with multiple
+fields, and many other valid but challenging ideas (documented in the "Future
+possibilities" section). This RFC defines a *simple* mechanism for defining one
+common type of niche, while leaving room for future extension.
+
+Defining a niche mechanism allows libraries to build arbitrary types containing
+niches, and simplifies handling of space-efficient data structures in Rust
+without manual bit-twiddling.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When defining a structure containing exactly one field of a non-zero-sized type
+(non-ZST), you can attach a `niche` attribute on it to declare a specific value
+or range of values for that field as invalid. This promises the compiler that
+you will never store those values in that field, which allows the compiler to
+use those in-memory representations for different purposes, such as the
+representation of `None` in a containing `Option`.
+
+```rust
+use std::mem::size_of;
+
+#[niche(value = 42)]
+struct MeaninglessNumber(u64);
+
+assert_eq!(size_of::<MeaninglessNumber>(), 8);
+assert_eq!(size_of::<Option<MeaninglessNumber>>(), 8);
+
+#[niche(range = 2..)]
+struct Bit(u8);
+
+assert_eq!(size_of::<Bit>(), 1);
+assert_eq!(size_of::<Option<Option<Option<Bit>>>>(), 1);
+```
+
+Constructing a structure with a niche value, or writing to the non-ZST field of
+such a structure, or obtaining a mutable reference to such a field, requires
+`unsafe` code. Causing a type with a niche to contain an invalid value (whether
+by construction, writing, or transmuting) results in undefined behavior.
+
+If a type `T` contains only a single niche value, `Option<T>` (and other enums
+isomorphic to it, with one variant containing `T` and one nullary variant) will
+use that value to represent `None` (the nullary variant). If such a `T` is
+additionally `repr(transparent)` or `repr(C)` or otherwise permitted in FFI,
+`Option<T>` will likewise be permitted in FFI, with the niche value mapping
+bidirectionally to `None` across the FFI boundary.
+
+If a type contains multiple niche values, Rust does not guarantee any
+particular mapping at this time, but may in the future.
+
+# Reference-level explanation
```

Comment:
fix: The reference-level section currently contains no mention of `unsafe`.  The guide-level and future possibilities section say that unsafe is required for some things, so the reference section should describe those rules precisely.

Reply:
@scottmcm I've now covered that in detail in the reference-level section.